reference/test/preconditioner/ilu.cpp

```diff
@@ -0,0 +1,317 @@
+                          .on(exec),
+                      gko::stop::ResidualNormReduction<>::build()
+                          .with_reduction_factor(1e-15)
+                          .on(exec))
+                  .on(exec)),
+          ilu_pre_factory(ilu_prec_type::build()
+                              .with_l_solver_factory(l_factory)
+                              .with_u_solver_factory(u_factory)
+                              .on(exec)),
+          ilu_rev_pre_factory(ilu_rev_prec_type::build()
+                                  .with_l_solver_factory(l_factory)
+                                  .with_u_solver_factory(u_factory)
+                                  .on(exec))
+    {}
+
+    std::shared_ptr<const gko::Executor> exec;
+    std::shared_ptr<Mtx> mtx;
+    std::shared_ptr<Mtx> l_factor;
+    std::shared_ptr<Mtx> u_factor;
+    std::shared_ptr<l_solver_type::Factory> l_factory;
+    std::shared_ptr<u_solver_type::Factory> u_factory;
+    std::shared_ptr<ilu_prec_type::Factory> ilu_pre_factory;
+    std::shared_ptr<ilu_rev_prec_type::Factory> ilu_rev_pre_factory;
+};
+
+
+TEST_F(Ilu, KnowsItsExecutor)
+{
+    auto ilu_factory = ilu_prec_type::build().on(exec);
+
+    ASSERT_EQ(ilu_factory->get_executor(), exec);
+}
+
+
+TEST_F(Ilu, CanSetLSolverFactory)
+{
+    auto ilu_factory =
+        ilu_prec_type::build().with_l_solver_factory(l_factory).on(exec);
+
+    ASSERT_EQ(ilu_factory->get_parameters().l_solver_factory, l_factory);
+}
+
+
+TEST_F(Ilu, CanSetUSolverFactory)
+{
+    auto ilu_factory =
+        ilu_prec_type::build().with_u_solver_factory(u_factory).on(exec);
+
+    ASSERT_EQ(ilu_factory->get_parameters().u_solver_factory, u_factory);
+}
+
+
+TEST_F(Ilu, BuildsDefaultWithoutThrowing)
+{
+    auto ilu_pre_default_factory = ilu_prec_type::build().on(exec);
+    auto par_ilu_fact =
+        gko::factorization::ParIlu<value_type>::build().on(exec);
+    auto par_ilu = par_ilu_fact->generate(mtx);
+
+    ASSERT_NO_THROW(ilu_pre_default_factory->generate(gko::share(par_ilu)));
+}
+
+
+TEST_F(Ilu, BuildsCustomWithoutThrowing)
+{
+    auto par_ilu_fact =
+        gko::factorization::ParIlu<value_type>::build().on(exec);
+    auto par_ilu = par_ilu_fact->generate(mtx);
+
+    ASSERT_NO_THROW(ilu_pre_factory->generate(gko::share(par_ilu)));
+}
+
+
+TEST_F(Ilu, ThrowOnWrongInput)
+{
+    ASSERT_THROW(ilu_pre_factory->generate(l_factor), gko::NotSupported);
+}
+
+
+TEST_F(Ilu, SolvesDefaultSingleRhsWithParIlu)
+{
+    const auto b = gko::initialize<Mtx>({1.0, 3.0, 6.0}, exec);
+    auto x = Mtx::create(exec, gko::dim<2>{3, 1});
+    auto par_ilu_fact =
+        gko::factorization::ParIlu<value_type>::build().on(exec);
+    auto par_ilu = par_ilu_fact->generate(mtx);
+
+    auto preconditioner =
+        default_ilu_prec_type::build().on(exec)->generate(gko::share(par_ilu));
+    preconditioner->apply(b.get(), x.get());
+
+    // Since it is a preconditioner, the default solve is not accurate
```

Comment:
In this case, it is exact :)

Reply:
That is correct, I will change it after I rebase it. I needed a lower precision, so the three iterations and the rel. res. of 1e-4 would pass.